components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,346 @@
+        return Err(format!("{} is not inside the base site directory {}", full_path.display(), base_path.display()).into());
+    }
+    return read_file(&full_path)
+        .map_err(|e| format!("`load_data`: error {} loading file {}", full_path.to_str().unwrap(), e).into());
+}
+
+fn get_output_format_from_args(args: &HashMap<String, Value>, provided_argument: &ProvidedArgument) -> Result<String> {
+    let format_arg = optional_arg!(
+        String,
+        args.get("format"),
+        "`load_data`: `format` needs to be an argument with a string value, being one of the supported `load_data` file types (csv, json, toml)"
+    );
+
+    if let Some(format) = format_arg {
+        return Ok(format);
+    }
+    return match provided_argument {
+        ProvidedArgument::PATH(path) => path.extension().map(|extension| extension.to_str().unwrap().to_string()).ok_or(format!("Could not determine format for {} from extension", path.display()).into()),
+        _ => Ok(String::from("plain"))
+    }
+}
+
+
+/// A global function to load data from a data file.
+/// Currently the supported formats are json, toml and csv
+pub fn make_load_data(content_path: PathBuf, base_path: PathBuf) -> GlobalFn {
+    let mut headers = header::HeaderMap::new();
+    headers.insert(header::USER_AGENT, format!("{} {}", crate_name!(), crate_version!()).parse().unwrap());
+    let client = Arc::new(Mutex::new(Client::builder().build().expect("reqwest client build")));
+    let result_cache: Arc<Mutex<HashMap<String, Value>>> = Arc::new(Mutex::new(HashMap::new()));
+    Box::new(move |args| -> Result<Value> {
+        let provided_argument = get_data_from_args(&content_path, &args)?;
+
+        let file_format = get_output_format_from_args(&args, &provided_argument)?;
+
+        let cache_key = get_cache_key(&provided_argument, &file_format);
+
+        let mut cache = result_cache.lock().expect("result cache lock");
+        let response_client = client.lock().expect("response client lock");
+        if let Some(cached_result) = cache.get(&cache_key) {
+            return Ok(cached_result.clone());
+        }
+
+        let data = match provided_argument {
+            ProvidedArgument::PATH(path) => read_data_file(&base_path, path),
+            ProvidedArgument::URL(url) => {
+                let mut response = response_client.get(url.as_str()).send().and_then(|res| res.error_for_status()).map_err(|e| format!("Failed to request {}: {}", url, e.status().expect("response status")))?;
+                response.text().map_err(|e| format!("Failed to parse response from {}: {:?}", url, e).into())
+            },
+        }?;
+
+        let result_value: Result<Value> = match file_format.as_str() {
+            "toml" => load_toml(data),
+            "csv" => load_csv(data),
+            "json" => load_json(data),
+            "plain" => to_value(data).map_err(|e| e.into()),
```

Comment:
I'm not sure about `plain`, what is it even loading?

Reply:
`plain` is just the format which returns the data raw, with no additional processing. Either the request body or the file contents.

Reply:
There probably are some use cases out there, and seeing as it was _basically free_ to implement, thought I might as well. It's probably something which may be critical to someone out there

Reply:
Just a heads up - I actually would make use of `plain` to load CSS and shove it inline at build-time, or to load a base64-encoded value that's ugly in code but needs to exist. They're rare use cases but they do exist.